### PR TITLE
feat(rest): Added new endpoints to update attachment info of Project, Component, Release

### LINKED ISSUE #1189: REST API: allow modification of attachment meta info (checkStatus)
### Description
(from sw360 user)

The REST endpoint for attachments currently only allows GET, so it's not possible to modify attachment information.
I'd be especially interested to automatically set "checkStatus", "checkedBy" and "checkedComment" information to confirm that an attachment was verified, e.g. when checking existing source files against known good ones.

### PR DESCRIPTION
**Pull request by JaideepPalit**
> Added new endpoints to update attachment info of Project, Component, Release.
> * Which issue is this pull request belonging to and how is it solving it? (*#1189*)
> * Did you add or update any new dependencies that are required for your change? - No

Issue: closes #1189 

http://localhost:8080/resource/docs/api-guide.html#resources-project-attachment-info-update
Example Request -
```
curl 'https://sw360.org/api/projects/98745/attachment/1234' -i -X PATCH 
-H 'Content-Type: application/hal+json' -H 'Authorization: Bearer **********' -H 'Accept: application/hal+json' 
-d '{
  "attachmentType" : "BINARY",
  "createdComment" : "Created Comment",
  "checkedComment" : "Checked Comment",
  "checkStatus" : "ACCEPTED"
}'
```
> http://localhost:8080/resource/docs/api-guide.html#resources-component-attachment-info-update
Example Request -
```
curl 'https://sw360.org/api/components/98745/attachment/1234' -i -X PATCH 
-H 'Content-Type: application/hal+json' -H 'Authorization: Bearer *******' -H 'Accept: application/hal+json' 
-d '{
  "attachmentType" : "BINARY",
  "createdComment" : "Created Comment",
  "checkedComment" : "Checked Comment",
  "checkStatus" : "ACCEPTED"
}'
```
>http://localhost:8080/resource/docs/api-guide.html#resources-release-attachment-info-update
Example Request -
```
curl 'https://sw360.org/api/releases/98745/attachment/1234' -i -X PATCH 
-H 'Content-Type: application/hal+json' -H 'Authorization: Bearer *************' -H 'Accept: application/hal+json' 
-d '{
  "attachmentType" : "BINARY",
  "createdComment" : "Created Comment",
  "checkedComment" : "Checked Comment",
  "checkStatus" : "ACCEPTED"
}'
```

> Have you implemented any additional tests? - No

### Checklist
Must:
- [X] All related issues are referenced in commit messages and in PR

Signed-off-by: Jaideep Palit <jaideep.palit@siemens.com>

### CODE REVIEW

**Review by akapti — CHANGES_REQUESTED** [review]
Reviewed the code, suggested some minor changes.
Tested the changes, and all the 3 endpoints working as expected.

File: rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/ProjectSpecTest.java [hunk 1]
```diff
@@ -142,6 +153,16 @@ public void before() throws TException, IOException {
...
+        setOfAttachment.add(att1);
+        Project projectForAtt = new Project();
+        projectForAtt.setAttachments(setOfAttachment);
+        projectForAtt.setId("98745");
+        projectForAtt.setName("Test Project");
+        projectForAtt.setProjectType(ProjectType.PRODUCT);
+        projectForAtt.setVersion("1");
+        projectForAtt.setCreatedOn("2021-04-27");
+        projectForAtt.setCreatedBy("admin@sw360.org");
+
```
[the remark this answers is not in the document]
Comment (JaideepPalit) [reply rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/ProjectSpecTest.java]: removed

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java [hunk 2]
```diff
@@ -249,6 +253,48 @@ public AttachmentContent getAttachmentContent(String id) {
         return attachmentClient.getUsedAttachments(Source.projectId(projectId), null);
     }
 
+    public void updateAttachment(Attachment attachmentToUpdate, Attachment reqBodyAttachment, User user) {
```
[the remark this answers is not in the document]
Comment (JaideepPalit) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java]: changed to private

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java [hunk 3]
```diff
@@ -249,6 +253,48 @@ public AttachmentContent getAttachmentContent(String id) {
...
+            }
+        }
+    }
+
+    public Attachment updateAttachment(Set<Attachment> attachments, Attachment newData, String attachmentId, User user) {
+        if (CommonUtils.isNotEmpty(attachments)) {
+            Optional<Attachment> matchingAttachment = attachments.stream()
+                    .filter(att -> att.attachmentContentId.equals(attachmentId)).findFirst();
+            if (matchingAttachment.isPresent()) {
+                Attachment actualAtt=matchingAttachment.get();
```
[the remark this answers is not in the document]
Comment (JaideepPalit) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java]: fixed

**Comment by JaideepPalit** [comment]
Thanks for review and test.
Incorporated changes based on review comments

**Review by akapti — APPROVED** [review]
Code looks good.
Tested the PR again after changes, and is working fine.

**Comment by Lythenas** [comment]
Is there a reason that the patch url is different from the get url?

To get the attachment infos you need to use `https://sw360.org/api/attachments/1234`
but to patch the same infos you need to use `https://sw360.org/api/attachment/1234`. I think this is really inconvenient.

It would also be nice if attachment infos could be updated through the `self` link (i.e. `https://sw360.org/api/attachments/5c597ceca6a04445b188daf8490e93b3`)

**Comment by mcjaeger** [comment]
so, it is about the "s" ... I agree, but I am not sure how much trouble we cause will there be in case we change it

**Pull request merged**